### PR TITLE
Window manipulation - moving and resizing windows

### PR DESCRIPTION
**Pull request by Antikyth**
This is an early implementation of moving and resizing (floating) windows.  Currently, there are issues with this PR, primarily that far too many `ConfigureWindow` requests are sent to the X server when manipulating windows, causing serious performance concerns. A way of limiting the frequency of updating the window when manipulating windows would ease these performance concerns.
Additionally, the window manager should translate the coordinates stored for the current window manipulation if the window is reparented. It should probably also deny `ConfigureWindow` requests from other sources for the duration of the window manipulation, and handle situations like the window being manipulated requesting to be unmapped, for instance.

### CODE REVIEW

File: src/aquariwm.rs
Comment (KodiCraft) [inline src/aquariwm.rs:96]: Ending manipulations should always happen upon release of the manipulation key, this may need to be changed.

File: src/main.rs
Comment (KodiCraft) [inline src/main.rs:68]: Is this information written in another unmissable location? It seems like critical information

File: src/window_manipulation.rs
Comment (KodiCraft) [inline src/window_manipulation.rs]: Wouldn't these two functions make more sense as macros if possible? They may add a little unnecessary overhead for simple shorthands

**Review by KodiCraft — DISMISSED** [review]
This looks really solid. A couple of minor things might need changing but especially as an early implementation this looks perfectly good to me.

File: src/aquariwm.rs
Comment (Antikyth) [reply src/aquariwm.rs:96]: That's a good point. The super key should only be relevant for the button press. There should probably be a separate `GrabPointer` request sent when window manipulation starts, on the manipulated window, that will stop the pointer events going to the window. Don't quote me on the details though - I know there is pointer motion, button motion, grabbing pointer, grabbing buttons... but something along those lines.

Grabbing the buttons/pointer makes it so the events aren't sent to the window itself. These need to be set specifically during window manipulation because, as you say, they should still be grabbed if the Super key is released.

File: src/main.rs
Comment (Antikyth) [reply src/main.rs:68]: Not yet, but I am planning on writing it in CONTRIBUTING.md or somewhere like that - this just isn't the appropriate location for that information: it is only readable when looking at source code, as it is comments, and it is contained within a pretty much unrelated function.

File: src/window_manipulation.rs
Comment (Antikyth) [reply src/window_manipulation.rs:150]: Yes, you're correct. I'll make sure to change them to macros. I only did them as functions because I'm not used to macros yet and didn't think of it - my mistake.

File: src/main.rs
Comment (Antikyth) [reply src/main.rs:68]: It might be good to have in doc comments perhaps? Whether that would be crate level documentation or something, I'm not sure.  I don't know if that is appropriate or not, but as this is a binary crate, my thinking was that people reading _these_ docs are contributors.  It does make sense to me to put it in CONTRIBUTING.md, at least in addition, however - CONTRIBUTING.md is for contributing guidelines, which these seem to be.

File: src/aquariwm.rs
Comment (KodiCraft) [reply src/aquariwm.rs:96]: As manipulation can only be performed on one window at a time as far as I know, this seems to be a good approach. Grabbing buttons while a manipulation is occuring and waiting for their release, although some UX tests would probably be necessary for it to not feel unnatural

File: src/aquariwm.rs
Comment (Antikyth) [inline src/aquariwm.rs]: This still feels a little clunky to me. Is there a better way to do this?

File: src/main.rs
Comment (KodiCraft) [reply src/main.rs:68]: I believe that information could also be useful to a general user looking to debug something, so in my opinion CONTRIBUTING.md isn't the ideal location. Maybe a page in the component documentation about logging that would be linked to in CONTRIBUTING and in some eventual Troubleshooting section could be better.

**Review by Antikyth — COMMENTED** [review]
There are still changes that need to be made to this code before it is ready to be merged:
- As pointed out by @KodiCraft, the Super key should only be required to _begin_ window manipulation. After that, releasing the Super key should not end the window manipulation - pointer input and buttons should be grabbed on the window when window manipulation starts, and ungrabbed when it ends, so that they aren't passed through to the window itself.
- Window manipulation needs to send far less `ConfigureWindow` requests. Right now, _hundreds_ of requests can be sent in a single drag, and this is affecting performance significantly. Luckily, this will be a problem all window managers face, so there should be plenty of references to resolve this.

Once those are addressed, I'll review the code again.

File: src/window_manipulation.rs
Comment (Antikyth) [reply src/window_manipulation.rs]: I don't think they would make more sense as macros, _but_ they would make sense as `const fn`s. As far as I can tell, that's the same kind of idea - `const fn`s are replaced at compile time, similar to macros but they aren't targeting syntax directly. `std`'s `Option` actually has pretty much the exact same thing as `const fn`s:
```rust
pub const fn is_some(&self) -> bool {
    matches!(*self, Some(_))
}
```
So I will change them to `const fn`.

File: src/aquariwm.rs
Comment (Antikyth) [reply src/aquariwm.rs:96]: > As manipulation can only be performed on one window at a time as far as I know, this seems to be a good approach. Grabbing buttons while a manipulation is occuring and waiting for their release, although some UX tests would probably be necessary for it to not feel unnatural

What do you mean exactly (re: unnatural)? I believe that's how pretty much all window managers do it - the buttons are grabbed because the intention is definitely not to interact with the window. I can imagine that another way to do it would be to make releasing Super cancel the window manipulation instead, though, but I think it makes sense that once you're dragging a window, you still want to be manipulating it until the mouse buttons are released. We could look at how the controls generally work in other window managers - might be bikeshedding though, I know `bspwm` works like the first suggestion where Super only matters to start the window manipulation, and its a small issue overall.

File: src/window_manipulation.rs
Comment (Antikyth) [inline src/window_manipulation.rs:258]: Might want to expand the doc comment here to cover how the manipulation is applied, and the exact effects of that (and the specifics for both moving and resizing)?

File: src/window_manipulation.rs
Comment (Antikyth) [inline src/window_manipulation.rs:316]: Is there a better way to do this (like a builtin way or something)? If not, should this be moved to a separate utility function, or is it fine where it is?

File: src/window_manipulation.rs
Comment (Antikyth) [inline src/window_manipulation.rs]: As the TODOs say here, these should definitely be moved out into a utility module, along with `crate::init_window`.

File: src/window_manipulation.rs
Comment (Antikyth) [inline src/window_manipulation.rs]: Can these be `const fn`s?

File: src/main.rs
Comment (Antikyth) [inline src/main.rs]: If changing to not require Super to be continually held, this grab should only be for `BUTTON_PRESS`. After receiving a button press event and starting window manipulation, an additional button grab should be made for the `BUTTON_RELEASE` and `BUTTON_MOTION` masks.

File: src/main.rs
Comment (Antikyth) [reply src/main.rs:68]: > I believe that information could also be useful to a general user looking to debug something, so in my opinion CONTRIBUTING.md isn't the ideal location. Maybe a page in the component documentation about logging that would be linked to in CONTRIBUTING and in some eventual Troubleshooting section could be better.

I see, that makes sense. I'd say in that case it should be 'duplicated': two descriptions, one more appropriate for contributors, the other for users troubleshooting?

File: src/window_manipulation.rs
Comment (Antikyth) [reply src/window_manipulation.rs]: Yes, they can.

File: src/window_manipulation.rs
Comment (Antikyth) [reply src/window_manipulation.rs]: Done.

**Pull request merged**
